Skeleton spawner

Dtinpa opened the pull request.

Dtinpa committed:
Added Skeleton Spawn Manager and Skeleton Spawner prefabs and scripts.

Skeleton Spawner:
- Uses overlapSphere to get colliders the Vector3 hits
- If there are colliders whose tags aren't excluded from this detection, then generate a new position
- Spawn rate increases for the closest spawner
- each spawner has a max amount of enemeies it spawns for easier game balance

Spawn Manager:
- Keeps track of all spawners in the game
- Checks each spawner to see how many enemies its spawned
- if the max global skeletons have been reached, turn off all spawners
- when the num of skeletons is less than the glocal max, and the spawners were previously deactivated, then reactivate them

Dtinpa committed:
fixed skeleton spawner since the spawnPos.z was using the xCollisionOffset instead of the zCollisionOffset

brandoncimino commented on Assets/Code/Runtime/SkeletonSpawner.cs:
```diff
@@ -0,0 +1,123 @@
+ 
+    private int skeletonCount = 0;
+
+    //if the global spawn limit has been reached, stops spawning enemies entirely
```
```suggestion
    /// if the global spawn limit has been reached, stops spawning enemies entirely
```

brandoncimino commented on Assets/Code/Runtime/SkeletonSpawner.cs:
```diff
@@ -0,0 +1,123 @@
+
+    //if the global spawn limit has been reached, stops spawning enemies entirely
+    private bool continueSpawning = true;
+    //if we tried to spawn something but it was still colliding, then skip instantiating so the game doesn't freeze
```
```suggestion
    /// if we tried to spawn something but it was still colliding, then skip instantiating so the game doesn't freeze
```

brandoncimino commented on Assets/Code/Runtime/SkeletonSpawner.cs:
```diff
@@ -0,0 +1,123 @@
+    private bool continueSpawning = true;
+    //if we tried to spawn something but it was still colliding, then skip instantiating so the game doesn't freeze
+    private bool couldSpawn = true;
+    //if this spawner is the closest to the player, then speed up the spawn rate
```
```suggestion
    /// if this spawner is the closest to the player, then speed up the spawn rate
```

Dtinpa committed:
committing meta files

brandoncimino commented on Assets/Code/Runtime/SkeletonSpawner.cs:
```diff
@@ -0,0 +1,123 @@
+        StartCoroutine(spawnEnemy(enemy, interval, enemyMaxSpawn));
+    }
+
+    private Vector3 determineSpawnLocation()
```
```suggestion
        private static RaycastHit HitTerrain(float x, float z) {
        var rayOrigin = new Vector3(x, 0, z);
        return Physics.Raycast(rayOrigin, Vector3.down, out var hit) ? hit : throw new InvalidOperationException($"Point (x: {x}, z: {z}) isn't in this terrestrial sphere!");
    }

    private Vector3? determineSpawnLocation() {
        var spawnX   = Random.Range(xAxisMaxRange, xAxisMaxRange);
        var spawnZ   = Random.Range(zAxisMaxRange, zAxisMaxRange);
        var position = transform.position;
        var spawnPos = HitTerrain(position.x + spawnX, position.z + spawnZ).point;

        const int maxLoops = 5;
        for (int i = 0; i < maxLoops; i++) {
            if (spawnIsColliding(spawnPos) == false) {
                return spawnPos;
            }
            
            spawnPos.x = (this.transform.position.x > spawnPos.x) ? spawnPos.x + xCollisionOffset : spawnPos.x - xCollisionOffset;
            spawnPos.z = (this.transform.position.z > spawnPos.z) ? spawnPos.z + zCollisionOffset : spawnPos.z - zCollisionOffset;
        }

        return null;
    }
```

Erichiller21 committed:
Update skull_island_v1.fbx

Erichiller21 committed:
Merge branch 'main' of https://github.com/brandoncimino/ludum-dare-51

Dtinpa committed:
Merge branch 'main' into skeleton-spawner

chezmeister0 committed:
skeleton boi

chezmeister0 committed:
Merge branch 'main' of https://github.com/brandoncimino/ludum-dare-51

chezmeister0 committed:
Create skeleton_mdl_v001.fbx

chezmeister0 committed:
Delete skeleton_mdl_v001.fbx

Dtinpa committed:
Merge branch 'main' into skeleton-spawner

chezmeister0 committed:
Update skeleton_mdl_v001.fbx

Dtinpa committed:
Merge branch 'main' into skeleton-spawner

Erichiller21 committed:
Create skull_island_v2.fbx

Erichiller21 committed:
Merge branch 'main' of https://github.com/brandoncimino/ludum-dare-51

Erichiller21 committed:
Delete skull_island_v2.fbx

Erichiller21 committed:
Create skull_island_v2.fbx

GuyWithPasta committed:
Added explosion sfx (first pass)

GuyWithPasta committed:
Merge remote-tracking branch 'origin/main'

Erichiller21 committed:
Create skull_island_v3.fbx

Erichiller21 committed:
Merge branch 'main' of https://github.com/brandoncimino/ludum-dare-51

chezmeister0 committed:
new bomb rig

chezmeister0 committed:
Merge branch 'main' of https://github.com/brandoncimino/ludum-dare-51

chezmeister0 committed:
rig fix

chezmeister0 committed:
color fix

chezmeister0 committed:
again!

brandoncimino committed:
Update Assets/Code/Runtime/SkeletonSpawner.cs

brandoncimino committed:
Update Assets/Code/Runtime/SkeletonSpawner.cs

brandoncimino committed:
Update Assets/Code/Runtime/SkeletonSpawner.cs

chezmeister0 committed:
palm tree!

GuyWithPasta committed:
Added GolfBomb Prefab (with wiggly wick action)

GuyWithPasta committed:
Merge remote-tracking branch 'origin/main'

GuyWithPasta committed:
Added Explosion sfx when Bomb Explodes

GuyWithPasta committed:
Changed IBlastable's method name to Blasted

Dtinpa committed:
Merge branch 'skeleton-spawner' of https://github.com/brandoncimino/ludum-dare-51 into skeleton-spawner

Dtinpa committed:
Update Assets/Code/Runtime/SkeletonSpawner.cs

added raycasting for more accurate spawning, cleaned up some more portions of code

Co-authored-by: Brandon Cimino <brandon.cimino@gmail.com>

Dtinpa committed:
Merge branch 'skeleton-spawner' of https://github.com/brandoncimino/ludum-dare-51 into skeleton-spawner

Dtinpa committed:
Merge branch 'main' into skeleton-spawner

Dtinpa committed:
changd spawning to happen under the SpawnManager, now it spawns based on a time interval.  Closest spawners get priority to spawn skeletons

Dtinpa commented:
Does this look good to merge in?
